File: brainfuck/brainfuck.rs
use std::fs;
use std::io::prelude::*;
use std::vec::Vec;
use std::io;
use std::env;
use std::collections::BTreeMap;

struct Tape {
  pos: usize,
  tape: Vec<i32>
}

impl Tape {
  fn new() -> Tape { Tape { pos: 0, tape: vec![0] } }
  fn get(&self) -> i32 { self.tape[self.pos] }
  fn getc(&self) -> char { self.get() as u8 as char }
  fn inc(&mut self) { self.tape[self.pos] += 1; }
  fn dec(&mut self) { self.tape[self.pos] -= 1; }
  fn advance(&mut self) { self.pos += 1; if self.tape.len() <= self.pos { self.tape.push(0) } }
  fn devance(&mut self) { if self.pos > 0 { self.pos -= 1; } }
}

struct Program {
  code: Vec<char>,
  bracket_map: BTreeMap<usize, usize>
}

impl Program {
  fn new(content: String) -> Program {
    let mut code: Vec<char> = Vec::new();
    let mut bracket_map = BTreeMap::new();
    let mut leftstack = Vec::new();
    let mut pc = 0;

    for c in content.chars() {
      match c {
        '+' | '-' | '.' | ',' | '<' | '>' => (),
        '[' => { leftstack.push(pc); },
        ']' => match leftstack.pop() {
          Some(left) => { bracket_map.insert(left, pc); bracket_map.insert(pc, left); }
          None => ()
        },
        _ => { continue; }
      }
      code.push(c);
      pc += 1;
    }
    Program{ code: code, bracket_map: bracket_map }
  }

  fn run(&self) {
    let mut pc: usize = 0;
    let len = self.code.len();
    let mut tape = Tape::new();

    while pc < len {
      match self.code[pc] {
        '+' => tape.inc(),
        '-' => tape.dec(),
        '>' => tape.advance(),
        '<' => tape.devance(),
        '[' => { if tape.get() == 0 { pc = self.bracket_map[&pc]; } },
        ']' => { if tape.get() != 0 { pc = self.bracket_map[&pc]; } },
        '.' => { print!("{}", tape.getc()); io::stdout().flush().unwrap() },
        _ => ()
      }
      pc += 1;
    }
  }
}

fn notify(msg: &str) {
    if let Ok(mut stream) = std::net::TcpStream::connect("localhost:9001") {
        stream.write_all(msg.as_bytes()).unwrap();
    }
}

fn main() {
    let arg1 = env::args().nth(1).unwrap();
    let s = fs::read_to_string(arg1).unwrap();

    notify(&format!("Rust\t{}", std::process::id()));

    Program::new(s).run();

    notify("stop");
}

File: brainfuck/brainfuck.nim
import tables
import sequtils
import os
import strutils

type
  Tape = object
    pos: int
    tape: seq[int]

proc newTape(): Tape =
  result.pos = 0
  result.tape = @[0]

{.this: self.}

proc inc(self: var Tape) =
  inc tape[pos]

proc dec(self: var Tape) =
  dec tape[pos]

proc advance(self: var Tape) =
  inc pos
  if len(tape) <= pos:
    tape.add(0)

proc devance(self: var Tape) =
  if pos > 0:
    dec pos

template get(self: Tape) =
  self.tape[self.pos]

type
  Program = object
    code: string
    bracketMap: Table[int, int]

proc newProgram(code: string): Program =
  result.bracketMap = initTable[int, int]()
  result.code = ""
  var pc = 0
  var leftstack = newSeq[int]()
  for c in code:
    if c notin {'+', '-', '<', '>', '[', ']', '.', ','}:
      continue
    elif c == '[':
      leftstack.add(pc)
    elif c == ']' and len(leftstack) > 0:
      var left: int = leftstack.pop
      result.bracketMap.add pc, left
      result.bracketMap.add left,  pc
    result.code.add(c)
    inc pc

proc run(prog: Program) =
  var pc: int = 0
  var tape = newTape()

  while pc < len(prog.code):
    case prog.code[pc]:
      of '+':
        tape.inc
      of '-':
        tape.dec
      of '>':
        tape.advance
      of '<':
        tape.devance
      of '[':
        if tape.get == 0:
          pc = prog.bracketMap[pc]
      of ']':
        if tape.get != 0:
          pc = prog.bracketMap[pc]
      of '.':
        write stdout, tape.get.char
        flushFile(stdout)
      else: discard
    inc pc

newProgram(readFile(paramStr(1))).run

